### ISSUE TITLE
GA4 - ecommerce dataLayer implementation

### ISSUE THREAD

**Issue by danielfmiranda** [MozillaFoundation/donate-wagtail#1671]
### Description: 

Implement GA4 - ecommerce dataLayer using the below code. 

### Context

This event and code is necessary to dynamically track donations made via the donation page. The code pushes the donation data into the dataLayer, which is then read by Google Tag Manager and sent to Google Analytics 4.

Must run when a user completes a donation.

### dataLayer code:

```
// Mozilla Foundation + Torchbox - 2022
// Ecommerce donation tracking for Google Analytics 4

window.dataLayer = window.dataLayer || [];
window.dataLayer.push({
  event: 'purchase',
  ecommerce: {
    currency: 'GBP', // 3-letter code for the currency in which the donation was made
    value: 9.99, // donation value in float point format
    transaction_id: ""T_12345"", // Transaction ID from payment processor
    items: [
        {
         item_name: ""regular donation"", // either 'regular donation' or 'single donation'
         currency: ""GBP"", // 3-letter code for the currency in which the donation was made
         index: 0, // keep at 0
         item_brand: ""credit card"", // name of payment method. either 'credit card' or 'paypal'
         item_category: ""donation"",
         item_variant: ""gift aid: yes"", // status of gift aid. either 'gift aid: yes' or 'gift aid: no'
         price: 9.99, // donation amount
         quantity: 1 // always ""1"". needs to be an integer
       }
      ],
  }
});
```

**Comment by cdanfon** [comment]
@fessehaye 

Can we get the GTM code added to http://donate-wagtail.**mofostaging**.net/ too? 

We'd need it to run test donations so we don't have to use an actual credit card on live. Code is below, to be added up high in the <head> element. Thanks!


<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'[https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f)](https://www.googletagmanager.com/gtm.js?id=%27+i+dl;f.parentNode.insertBefore(j,f));
})(window,document,'script','dataLayer','GTM-TCB2S5V');</script>

**Comment by cdanfon** [comment]
Hey @mtdenton @danielfmiranda @fessehaye 

The TBX digital team have just confirmed that this doesn't look like firing on production or staging - reopening the ticket. 

Can we please get the following code added to donate both on prod and staging? 

"// Mozilla Foundation + Torchbox - 2022
// Ecommerce donation tracking for Google Analytics 4

window.dataLayer = window.dataLayer || [];
window.dataLayer.push({
  event: 'purchase',
  ecommerce: {
    currency: 'GBP', // 3-letter code for the currency in which the donation was made
    value: 9.99, // donation value in float point format
    transaction_id: ""T_12345"", // Transaction ID from payment processor
    items: [
        {
         item_name: ""regular donation"", // either 'regular donation' or 'single donation'
         currency: ""GBP"", // 3-letter code for the currency in which the donation was made
         index: 0, // keep at 0
         item_brand: ""credit card"", // name of payment method. either 'credit card' or 'paypal'
         item_category: ""donation"",
         item_variant: ""gift aid: yes"", // status of gift aid. either 'gift aid: yes' or 'gift aid: no'
         price: 9.99, // donation amount
         quantity: 1 // always ""1"". needs to be an integer
       }
      ],
  }
});"

**Comment by cdanfon** [comment]
Hey @mtdenton @danielfmiranda @fessehaye

The TBX digital team have just confirmed that this doesn't look like firing on production or staging - reopening the ticket.

Can we please get the following code added to donate both on prod and staging?

```
"// Mozilla Foundation + Torchbox - 2022
// Ecommerce donation tracking for Google Analytics 4

window.dataLayer = window.dataLayer || [];
window.dataLayer.push({
  event: 'purchase',
  ecommerce: {
    currency: 'GBP', // 3-letter code for the currency in which the donation was made
    value: 9.99, // donation value in float point format
    transaction_id: ""T_12345"", // Transaction ID from payment processor
    items: [
        {
         item_name: ""regular donation"", // either 'regular donation' or 'single donation'
         currency: ""GBP"", // 3-letter code for the currency in which the donation was made
         index: 0, // keep at 0
         item_brand: ""credit card"", // name of payment method. either 'credit card' or 'paypal'
         item_category: ""donation"",
         item_variant: ""gift aid: yes"", // status of gift aid. either 'gift aid: yes' or 'gift aid: no'
         price: 9.99, // donation amount
         quantity: 1 // always ""1"". needs to be an integer
       }
      ],
  }
});
```

**Comment by danielfmiranda** [comment]
Hi @cdanfon, while working on this ticket I tried visiting the GTM dashboard to test out any changes, however, I cannot seem to even find an instance of the donate site on the dash board.

I was wondering if you could please post a link to the URL that the torchbox team was using for testing? 

Thanks in advance!

**Comment by mtdenton** [comment]
@cdanfon I sent this to @danielfmiranda via Slack

**Comment by danielfmiranda** [comment]
Hi @cdanfon!

I have a PR for this ticket pretty much ready to go, however, before I submit it for review I was wondering if you could help me answer these following questions:

**Question 1:** 
Are we only wanting to track payments made on the foundation's donate site, and not on the thunderbird site?

**Question 2:**
I was curious what the expectation was around tracking subscription payments (the payments that are automatically made one month after the initial "monthly" donation)? 

If I understand correctly, it appears that these are made by the Braintree/Paypal backend, and not our code.

But we do track the initial donation that **begins** the subscription, and update the pushed datalayer object to mark the donation "recurring" as requested.


Thanks in advance!

**Comment by cdanfon** [comment]
Thanks @danielfmiranda I'm checking with our Digital Marketing colleague here at TBX

**Comment by danielfmiranda** [comment]
Hi @cdanfon!

There are no more outstanding changes on this ticket, it is just awaiting review to be merged into staging, thanks!

**Comment by cdanfon** [comment]
Hey @danielfmiranda who are you waiting on review from?

**Comment by danielfmiranda** [comment]
This ticket is now on staging and ready to be reviewed, however, heroku is experiencing some issues which are causing donations on the staging site to return 500 errors. 

Going to continue monitoring heroku status, and will post a comment when the staging site is working again!

**Comment by danielfmiranda** [comment]
Heroku status is now fixed, however, it looks like we are experiencing a memory limit on the donate-staging app. There is a team discussion going on in slack to decide how we want to resolve this, and I will move this back to "ready for QA" once everything is set up and working correctly!

**Comment by cdanfon** [comment]
Thanks @danielfmiranda keep me posted please :)

**Comment by danielfmiranda** [comment]
**[Ready for testing]**
Hi @cdanfon Testing has been unblocked and is now ready on staging!

**Comment by cdanfon** [comment]
Thanks @danielfmiranda I've asked the TBX folk to test it on staging.

**Comment by danielfmiranda** [comment]
Just tested this on staging and I can confirm that the correct events are being pushed to the datalayer for the following:

- 1 time Donation (Credit card)
- 1 time donation - upsell (credit card)
- 1 time Donation (Paypal)
- 1 time donation - upsell (Paypal)
- Monthly donation (credit card)
- Monthly donation (paypal)

**Comment by bheasman** [comment]
Thanks @danielfmiranda ! This has been tested our end too and looking good. @danielfmiranda has this been pushed to production yet?
